File: src/tools/documents.ts
import { z } from "zod";
import { IdSchema, JsonObjectSchema } from "../schemas/projeqtor-types.js";
import { safeTool } from "../utils/error-handler.js";
import { type ToolContext } from "./common.js";

export function registerDocumentTools(server: any, ctx: ToolContext): void {
  server.registerTool("list_documents", {
    title: "List project documents",
    description: "List Document objects linked to a project or product. Metadata only; binary upload/download is intentionally not handled by this REST wrapper.",
    inputSchema: { projectId: IdSchema.optional(), productId: IdSchema.optional() }
  }, (a: any) => safeTool(() => {
    const criteria = [
      ...(a.projectId ? [{ field: "idProject", operator: "=", value: a.projectId }] : []),
      ...(a.productId ? [{ field: "idProduct", operator: "=", value: a.productId }] : [])
    ];
    return criteria.length ? ctx.client.search("Document", criteria) : ctx.client.listAll("Document");
  }));

  server.registerTool("create_document_record", {
    title: "Create document metadata record",
    description: "Create a ProjeQtOr Document metadata record. Use for registering documents; actual binary attachment handling depends on ProjeQtOr instance endpoints/plugins.",
    inputSchema: { name: z.string().min(1), projectId: IdSchema.optional(), description: z.string().optional(), extra: JsonObjectSchema.optional() }
  }, (a: any) => safeTool(() => ctx.client.create("Document", { name: a.name, idProject: a.projectId, description: a.description, ...(a.extra ?? {}) })));
}

File: src/projeqtor_mcp_server/tools/documents.py
from __future__ import annotations

from typing import Any

from mcp.server.fastmcp import FastMCP

from projeqtor_mcp_server.client.projeqtor_api import ProjeQtOrApiClient, SearchCriterion
from projeqtor_mcp_server.tools.common import Id, merge_extra, safe


def register_document_tools(mcp: FastMCP, client: ProjeQtOrApiClient) -> None:
    """Register document metadata tools."""

    @mcp.tool(description="Lister les documents liés à un projet ou produit. Métadonnées uniquement.")
    async def list_documents(project_id: Id | None = None, product_id: Id | None = None) -> Any:
        criteria = []
        if project_id is not None:
            criteria.append(SearchCriterion("idProject", project_id))
        if product_id is not None:
            criteria.append(SearchCriterion("idProduct", product_id))
        return await safe(client.search("Document", criteria) if criteria else client.list_all("Document"))

    @mcp.tool(description="Créer une fiche métadonnée Document. L'upload binaire dépend des endpoints/plugins de l'instance.")
    async def create_document_record(name: str, project_id: Id | None = None, description: str | None = None, extra: dict[str, Any] | None = None) -> Any:
        payload = merge_extra({"name": name, "idProject": project_id, "description": description}, extra)
        return await safe(client.create("Document", payload))
